### ISSUE TITLE
Level editor should not allow non-treatable diseass

### ISSUE THREAD

**Issue by bobcanthelpyou** [CorsixTH/CorsixTH#749]
When a patient is diagnosed as pregnant and try to leave the GP's office an error occurs and the patient get stuck at the rooms door.

![bildschirmfoto von 2015-06-09 20 46 16](https://cloud.githubusercontent.com/assets/4523264/8066610/d9c2b30e-0ee8-11e5-8469-599fd7ee0102.png)

![bildschirmfoto von 2015-06-09 20 16 33](https://cloud.githubusercontent.com/assets/4523264/8066284/16dfda02-0ee7-11e5-9d12-2a5c443c7287.png)

If i try to send the patient home the patient is stuck outside the office at the door.

![bildschirmfoto von 2015-06-09 20 16 42](https://cloud.githubusercontent.com/assets/4523264/8066291/1ff1a454-0ee7-11e5-885f-05dab33bcfb7.png)

```
Will not check for updates since this is the Trunk version.
An error has occured!
Almost anything can be the cause, but the detailed information below can help the developers find the source of the error.
Running: The timer handler.
A stack trace is included below, and the handler has been disconnected.
./Lua/humanoid_actions/seek_room.lua:104: room operating_theatre not available
stack traceback:
    [C]: in function 'assert'
    ./Lua/humanoid_actions/seek_room.lua:104: in function 'action_seek_room_no_treatment_room_found'
    ./Lua/humanoid_actions/seek_room.lua:284: in function <./Lua/humanoid_actions/seek_room.lua:201>
    ./Lua/entities/humanoid.lua:575: in function 'Humanoid_startAction'
    ./Lua/entities/humanoid.lua:667: in function 'finishAction'
    ./Lua/humanoid_actions/walk.lua:152: in function 'timer_function'
    ./Lua/entity.lua:218: in function 'tick'
    ./Lua/world.lua:1108: in function 'onTick'
    ./Lua/app.lua:876: in function 'dispatch'
    ./Lua/app.lua:777: in function <./Lua/app.lua:772>

Warning: No event handler for timer
```

Single Scenario: Avatar 
CorsixTH version: 0.40 and master branch

**Comment by TheCycoONE** [comment]
Could you post a save game?

**Comment by bobcanthelpyou** [comment]
Now it is the patient in the ward: 
http://filebin.ca/24gmB2gYGKZc

**Comment by MarkL1961** [comment]
This disease needs both a ward and operating theatre to be treated.  Usually you would get a message/fax telling you that one or other is missing and you are asked what you want to do  i.e. wait while you build the room, send home etc.
From the above error message it looks like there is no operating theatre (or is it that it is not staffed).  I can't check the save at the moment, but if that is the case, why has the previously mentioned fax not been received?  Is this just a trunk thing as it has been working fine for me with 0.40 and a copy of the trunk from about two months ago?

**Comment by bobcanthelpyou** [comment]
I could reproduce this bug with a trunk build of the master branch and the 0.40 release binary https://github.com/CorsixTH/CorsixTH/releases/tag/v0.40

The problem might be the Single Scenario Avatar, there is only one doctor available at all. (maybe the operating theatre is disabled for that scenario, but i don't know)

The error occurs if a GP's Office and a Ward is already available and the patient wants to leave the GP's Office or after you build a Ward after it is requested for the pregnant patient and the patient wants to leave the Ward.

**Comment by MarkL1961** [comment]
Sounds like it is specific to that level.  You would have thought that any diseases needing two doctors would have been disabled, so that perhaps could be part of the solution.
However, not sure if your Dr is a surgeon or not, but either way there should be a fax telling you that you can't treat that patient as the room does not exist and/or you have not the staff for it.
I've looked in the level file and there are other diseases in there that need the operating theatre!

**Comment by MarkL1961** [comment]
Running your save I can confirm your error.   I am not sure why it is not sending a fax for a decision as you clearly cannot cure the disease.
I am going to run some more test on this level

**Comment by Alberth289346** [comment]
I once reported a bug that the game sent people to the operating theatre without having staff for it, and got as reply that it was correct behaviour.

**Comment by MarkL1961** [comment]
I'd have to check it against TH, but I feel sure there was a fax to say you couldn't cure the disease, and the usual choices and that it would also show the red x in the casebook.  I know we don't have the casebook bit fully right, because it does not go back to the red x if you lose a surgeon or room, but I did think we had the fax right for this.

**Comment by MarkL1961** [comment]
@Alberth289346 I have been trying to remember the bug report you refer to and the only one I can recall is getting a fax for a decision where you don't have a room and cannot build it yet anyway.  Is that the one you mean or was there another one?

**Comment by Alberth289346** [comment]
https://code.google.com/p/corsix-th/issues/detail?id=1018&can=1&q=reporter%3AAlberth289346

**Comment by MarkL1961** [comment]
Yep, that was the one I was thinking of.  In your case you are getting a fax, this issue is you are not getting one.

**Comment by MarkL1961** [comment]
Played this map from start and each new disease generates a fax - disease discovered and if you cannot cure a fax- decision needed.  Then when pregnant patient arrives the game will crash as she exits the GP office.  Never got to other operable diseases, but assuming it will be the same.

**Comment by MarkL1961** [comment]
![fax3](https://cloud.githubusercontent.com/assets/4965487/8117265/bde4dfb2-107e-11e5-8c6a-d9daaa7dabd9.png)
![fax1](https://cloud.githubusercontent.com/assets/4965487/8117264/bde34f9e-107e-11e5-9736-739665d7689c.png)
![fax2](https://cloud.githubusercontent.com/assets/4965487/8117266/bde6f7fc-107e-11e5-998a-e6d4f997cd92.png)
In situations where you can build the operating theatre and haven't the faxes are generated.  If you can't build it as it is not available the game will error as above, then you get the first fax but not the  decision fax

**Comment by TheCycoONE** [comment]
Thanks for the full investigation @MarkL1961. So as I read it this is purely a LEVEL file issue.
Actionable items are to: a) Remove all surgical diseases from Avatar level
b) Create a new enhancement issue to scan for and reject or auto correct with warning such LEVEL files so that map makers don't accidentally make crashing levels.

**Comment by TheCycoONE** [comment]
Ok, then I misunderstood what you were saying.

The operating table isn't researchable, and you wouldn't be able to purchase it when editing the room: (https://github.com/CorsixTH/CorsixTH/blob/18e45548084068dffc0c01066356d2499eae8fc3/CorsixTH/Lua/dialogs/edit_room.lua#L510), and it is needed for the operating theatre - so I don't know how building an operating theatre would work.

Now to look from the other direction from the actual error, the assert that's failing is:
humanoid.world.available_rooms[room_type]

Now, if we look in world.lua available_rooms is populated in initRooms.  That function looks at each object_needed for a room. If that object is available from the start it says the room is available and it ends up in the list, if the object is available for the level it says the room is discoverable and it's added the list, if it's not the room is not added to the list. All this adds up to the operating theatre not being buildable in Avatar.

This is why the error happens in this level and why the fax isn't being displayed.

**Comment by Alberth289346** [comment]
Moved the self.available_rooms initialization out of initRooms (and renamed to getRooms).

**Comment by TheCycoONE** [comment]
+1

**Comment by TheCycoONE** [comment]
Oh, one more thing while were on this.  Can we call it getAvailableRooms to avoid confusion with World:rooms (which is all built rooms)

**Comment by Alberth289346** [comment]
done

**Comment by Alberth289346** [comment]
With commit:12d35845f8a60b5e8c24ba4db2d144cf79b33039 the game does not crash any more. Instead it removes diseases that cannot be treated.
To avoid such surprises for a level designer, the level editor should warn about it, or prevent this situation from happening at all.

**Comment by lewri** [comment]
May be resolved by PR #766
